### PR TITLE
ddl: truncate table by DDL operation.

### PR DESCRIPTION
**Pull request by coocood**
Currently, truncate table deletes rows in table in one large transaction,
too large transaction is very likely to fail.
This commit use DDL to do the work, it returns instantly, old data will be
deleted by background worker.

### CODE REVIEW

**Comment by coocood** [comment]
fixes https://github.com/pingcap/tidb/issues/1693

@zimulala @shenli @hanfei1991 PTAL

File: ddl/ddl_db_test.go [hunk 1]
```diff
@@ -571,3 +572,48 @@ func (s *testDBSuite) TestUpdateMultipleTable(c *C) {
...
+	tk.MustExec("truncate table t")
+
+	tk.MustExec("insert t values (3, 3), (4, 4)")
+	tk.MustQuery("select * from t").Check(testkit.Rows("3 3", "4 4"))
+
+	is = sessionctx.GetDomain(ctx).InfoSchema()
+	newTblInfo, err := is.TableByName(model.NewCIStr("test"), model.NewCIStr("t"))
+	c.Assert(newTblInfo.Meta().ID, Greater, oldTblID)
+
+	// verify that the old table data has been deleted by background worker.
```
Comment (shenli) [inline ddl/ddl_db_test.go]: Is there any chance that bg worker do not have enough time to delete the old table data?

File: ddl/bg_worker.go [hunk 2]
```diff
@@ -92,10 +92,13 @@ func (d *ddl) runBgJob(t *meta.Meta, job *model.Job) {
 		if job.State != model.JobCancelled {
 			log.Errorf("[ddl] run background job err %v", errors.ErrorStack(err))
 		}
-
 		job.Error = toTError(err)
 		job.ErrorCount++
+		return
 	}
+
+	job.SchemaState = model.StateNone
```
Comment (shenli) [inline ddl/bg_worker.go]: Is this a bug-fix?

File: ddl/table.go [hunk 3]
```diff
@@ -181,7 +155,46 @@ func (d *ddl) getTableInfo(t *meta.Meta, job *model.Job) (*model.TableInfo, erro
...
 }
 
-func (d *ddl) dropTableData(t table.Table) error {
-	err := d.delKeysWithPrefix(tablecodec.EncodeTablePrefix(t.Meta().ID), bgJobFlag)
+func (d *ddl) dropTableData(tableID int64) error {
+	err := d.delKeysWithPrefix(tablecodec.EncodeTablePrefix(tableID), bgJobFlag)
 	return errors.Trace(err)
 }
+
+func (d *ddl) onTruncateTable(t *meta.Meta, job *model.Job) error {
```
Comment (shenli) [inline ddl/table.go]: It is better to add some comments for the function.

**Comment by shenli** [comment]
Rest LGTM

File: ddl/ddl_db_test.go (hunk 1 above)
Comment (coocood) [inline ddl/ddl_db_test.go]: If very unlikely that data hasn't be deleted in 3 seconds.

File: ddl/bg_worker.go (hunk 2 above)
Comment (coocood) [inline ddl/bg_worker.go]: No, I just move the code out of `delRegorgTable` function.

**Comment by shenli** [comment]
@coocood Any update?

**Comment by shenli** [comment]
@zimulala PTAL

**Comment by shenli** [comment]
LGTM
@zimulala PTAL

File: ddl/ddl.go [hunk 4]
```diff
@@ -1083,6 +1084,31 @@ func (d *ddl) DropTable(ctx context.Context, ti ast.Ident) (err error) {
...
+		return errors.Trace(err)
+	}
+	job := &model.Job{
+		SchemaID: schema.ID,
+		TableID:  tb.Meta().ID,
+		Type:     model.ActionTruncateTable,
+		Args:     []interface{}{newTableID},
+	}
+	err = d.doDDLJob(ctx, job)
+	err = d.hook.OnChanged(err)
```
Comment (zimulala) [inline ddl/ddl.go]: s/hook.OnChanged/callHookOnChanged

File: ddl/ddl_db_test.go [hunk 5]
```diff
@@ -571,3 +572,48 @@ func (s *testDBSuite) TestUpdateMultipleTable(c *C) {
...
+	tablePrefix := tablecodec.EncodeTablePrefix(oldTblID)
+	hasOldTableData := true
+	for i := 0; i < 30; i++ {
+		err = kv.RunInNewTxn(store, false, func(txn kv.Transaction) error {
+			it, err1 := txn.Seek(tablePrefix)
+			if err1 != nil {
+				return err1
+			}
+			hasOldTableData = it.Key().HasPrefix(tablePrefix)
```
Comment (zimulala) [inline ddl/ddl_db_test.go]: Do we need to add the judgment of `Valid` ?

**Comment by zimulala** [comment]
LGTM

**Pull request merged**